backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/questionsubmission/QuestionSubmissionService.java

```diff
@@ -0,0 +1,278 @@
+        QuestionSubmission questionSubmission = getQuestionSubmission(questionSubmissionId);
+        Review.Status status = inReview ? Review.Status.IN_REVIEW : Review.Status.IN_REVISION;
+        updateQuestionStatus(status, questionSubmission.getQuestion().getId());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<QuestionSubmissionDto> getStudentQuestionSubmissions(Integer studentId, Integer courseExecutionId) {
+        return questionSubmissionRepository.findQuestionSubmissions(studentId, courseExecutionId).stream().map(QuestionSubmissionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<QuestionSubmissionDto> getCourseExecutionQuestionSubmissions(Integer courseExecutionId) {
+        return questionSubmissionRepository.findCourseExecutionQuestionSubmissions(courseExecutionId).stream().map(QuestionSubmissionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<ReviewDto> getQuestionSubmissionReviews(Integer questionSubmissionId) {
+        return reviewRepository.findQuestionSubmissionReviews(questionSubmissionId).stream().map(ReviewDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<UserQuestionSubmissionInfoDto> getAllStudentsQuestionSubmissionsInfo(Integer courseExecutionId) {
+        CourseExecution courseExecution = getCourseExecution(courseExecutionId);
+        Set<User> students = courseExecution.getStudents();
+
+        List<UserQuestionSubmissionInfoDto> userQuestionSubmissionInfoDtos = new ArrayList<>();
+
+        for (User student: students) {
+            List<QuestionSubmissionDto> questionSubmissions = student.getQuestionSubmissions().stream().map(QuestionSubmissionDto::new).collect(Collectors.toList());
+            userQuestionSubmissionInfoDtos.add(new UserQuestionSubmissionInfoDto(student, questionSubmissions));
+        }
+
+        userQuestionSubmissionInfoDtos.sort(UserQuestionSubmissionInfoDto.NumSubmissionsComparator);
+
+        return userQuestionSubmissionInfoDtos;
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void resetDemoQuestionSubmissions() {
+        questionSubmissionRepository.findCourseExecutionQuestionSubmissions(courseService.getDemoCourse().getCourseExecutionId())
+                .stream()
+                .skip(2)
```

Comment:
why this skip, this is a case where a comment is necessary

Reply:
it is not necessary, my bad